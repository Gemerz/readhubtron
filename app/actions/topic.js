// @flow
import { push } from 'react-router-redux';
// import type { TopicStateType } from '../reducers/topic';
import { checkHttpStatus, parseJSON } from '../utils';

type actionType = {
  type: string,
  payload: any
};
type TopicStateType = {
  topic: {
    collection: {
      length: number
    }
  }
};

type SettingStateType = {
  setting: {}
};
export const GET_TOPIC_INIT_REQUEST = 'GET_TOPIC_INIT_REQUEST';
export const GET_TOPIC_INIT_SUCCESS = 'GET_TOPIC_INIT_SUCCESS';
export const GET_TOPIC_INIT_FAILURE = 'GET_TOPIC_INIT_FAILURE';

export const GET_TOPIC_MORE_REQUEST = 'GET_TOPIC_MORE_REQUEST';
export const GET_TOPIC_MORE_SUCCESS = 'GET_TOPIC_MORE_SUCCESS';
export const GET_TOPIC_MORE_FAILURE = 'GET_TOPIC_MORE_FAILURE';

export const SET_TOPIC_LAST_CURSOR = 'SET_TOPIC_LAST_CURSOR';
export const SET_TOPIC_CURRENT_URL = 'SET_TOPIC_CURRENT_URL';
export const SET_TOPIC_TOTAL_URL = 'SET_TOPIC_TOTAL_URL';

export const CHECK_TOPIC_LATEST = 'CHECK_TOPIC_LATEST';


export function getTopicRequest() {
  return {
    type: GET_TOPIC_INIT_REQUEST,
    payload: { loading: true }
  };
}
export function getTopicSuccess(collection: any) {
  return {
    type: GET_TOPIC_INIT_SUCCESS,
    payload: { collection, loading: false }
  };
}


export function getTopicMoreRequest() {
  return {
    type: GET_TOPIC_MORE_REQUEST,
    payload: { moreLoading: true }
  };
}
export function getTopicMoreSuccess(collection: any) {
  return {
    type: GET_TOPIC_MORE_SUCCESS,
    payload: { collection, moreLoading: false }
  };
}
export function setTopiclastCursor(lastCursor: number) {
  return {
    type: SET_TOPIC_LAST_CURSOR,
    payload: { lastCursor }
  };
}
export function setTopicCurrentUrl(currentUrl: string) {
  return {
    type: SET_TOPIC_CURRENT_URL,
    payload: { currentUrl }
  };
}
export function setTopicTotalUrls(totalUrls: {}) {
  return {
    type: SET_TOPIC_TOTAL_URL,
    payload: { totalUrls }
  };
}
export function checkTopicLatest(count: number) {
  return {
    type: CHECK_TOPIC_LATEST,
    payload: { count }
  };
}


export function initTopic() {
  return (dispatch: (action: actionType) => void, getState: () => SettingStateType) => {
    const { setting } = getState();

    const API = 'https://api.readhub.me/topic?&pageSize=10';
    dispatch(getTopicRequest());
    fetch(API, {
      method: 'get',
      mode: 'cors',
      headers: {
        'Content-Length': ' application/json'
      }
    }).then(checkHttpStatus)
      .then(parseJSON)
      .then(response => {
        // dispatch(getTopic(response.data));
        dispatch(getTopicSuccess(response.data));
        if (response.data[0].order) {
          dispatch(setTopiclastCursor(response.data[0].order));
          const url = setting.moblieFirst ? response.data[0].newsArray[0].mobileUrl : response.data[0].newsArray[0].url;
          dispatch(setTopicCurrentUrl(url));
          dispatch(setTopicTotalUrls(response.data[0].newsArray));
        }
      });
  };
}

export function loadMoreTopic() {
  return (dispatch: (action: actionType) => void, getState: () => TopicStateType) => {
    const { topic } = getState();
    const nextPageCursor = topic.collection[topic.collection.length - 1].order;
    const API = `https://api.readhub.me/topic?&pageSize=10&lastCursor=${nextPageCursor}`;
    dispatch(getTopicMoreRequest());
    setTimeout(() => {
      fetch(API, {
        method: 'get',
        mode: 'cors',
        headers: {
          'Content-Length': ' application/json'
        }
      }).then(checkHttpStatus)
        .then(parseJSON)
        .then(response => {
          dispatch(getTopicMoreSuccess(response.data));
        });
    }, 500);
  };
}
export function fetchLatestCollection(lastCursor: number) {
  return (dispatch: (action: actionType) => void, getState: () => TopicStateType) => {
    const API = ` https://api.readhub.me/topic/newCount?latestCursor=${lastCursor}`;
    fetch(API, {
      method: 'get',
      mode: 'cors',
      headers: {
        'Content-Length': ' application/json'
      }
    }).then(checkHttpStatus)
      .then(parseJSON)
      .then(response => {
        dispatch(checkTopicLatest(response.count));
      });
  };
}
