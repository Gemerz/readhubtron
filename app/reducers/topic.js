// @flow
import {
  GET_TOPIC_INIT_REQUEST,
  GET_TOPIC_INIT_SUCCESS,
  GET_TOPIC_INIT_FAILURE,
  GET_TOPIC_MORE_REQUEST,
  GET_TOPIC_MORE_SUCCESS,
  GET_TOPIC_MORE_FAILURE,
  SET_TOPIC_LAST_CURSOR,
  SET_TOPIC_CURRENT_URL,
  SET_TOPIC_TOTAL_URL,
  CHECK_TOPIC_LATEST
} from '../actions/topic';

export type TopicStateType = {
  collection: any,
  lastCursor: number,
  loading: boolean,
  currentUrl: string,
  moreLoading: boolean,
  count: number,
  category: string,
  totalUrls: {}
};

type actionType = {
  type: string,
  payload: any
};
const initialState: TopicStateType = {
  collection: [],
  lastCursor: 0,
  loading: false,
  currentUrl: '',
  moreLoading: false,
  count: 0,
  category: 'topic',
  totalUrls: {}
};

export default function topic(state: TopicStateType = initialState, action: actionType) {
  const Collection = [];
  const MoreCollection = state.collection;
  switch (action.type) {
    // init list
    case GET_TOPIC_INIT_REQUEST:
      return {
        ...state,
        loading: action.payload.loading
      };
    case GET_TOPIC_INIT_SUCCESS:
      action.payload.collection.map((item) => Collection.push(item));
      return {
        ...state,
        collection: Collection,
        loading: action.payload.loading
      };
    case GET_TOPIC_INIT_FAILURE:
      return {
        ...state
      };
    // more list
    case GET_TOPIC_MORE_REQUEST:
      return {
        ...state,
        moreLoading: action.payload.moreLoading
      };
    case GET_TOPIC_MORE_SUCCESS:
      action.payload.collection.map((item) => MoreCollection.push(item));
      return {
        ...state,
        collection: MoreCollection,
        moreLoading: action.payload.moreLoading
      };
    case GET_TOPIC_MORE_FAILURE:
      return {
        ...state
      };
    case SET_TOPIC_LAST_CURSOR:
      return {
        ...state,
        lastCursor: action.payload.lastCursor,
      };

    case SET_TOPIC_CURRENT_URL:

      return {
        ...state,
        currentUrl: action.payload.currentUrl
      };
    case SET_TOPIC_TOTAL_URL:

      return {
        ...state,
        totalUrls: action.payload.totalUrls
      };
    // check lastest
    case CHECK_TOPIC_LATEST:
      return {
        ...state,
        count: action.payload.count,
      };

    default:
      return state;
  }
}
