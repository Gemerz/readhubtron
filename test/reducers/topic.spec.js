import topic from '../../app/reducers/topic';
// import { INCREMENT_COUNTER, DECREMENT_COUNTER } from '../../app/actions/counter';

describe('reducers', () => {
  describe('topic', () => {
    it('should handle initial state', () => {
      expect(topic(undefined, {})).toMatchSnapshot();
    });
    // it('should handle INCREMENT_COUNTER', () => {
    //   expect(counter(1, { type: INCREMENT_COUNTER })).toMatchSnapshot();
    // });
    // it('should handle DECREMENT_COUNTER', () => {
    //   expect(counter(1, { type: DECREMENT_COUNTER })).toMatchSnapshot();
    // });
    // it('should handle unknown action type', () => {
    //   expect(counter(1, { type: 'unknown' })).toMatchSnapshot();
    // });
  });
});
