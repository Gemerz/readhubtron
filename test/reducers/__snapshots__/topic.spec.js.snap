// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers topic should handle initial state 1`] = `
Object {
  "category": "topic",
  "collection": Array [],
  "count": 0,
  "currentUrl": "",
  "lastCursor": 0,
  "loading": false,
  "moreLoading": false,
  "totalUrls": Object {},
}
`;
